### PR TITLE
HAA0101 add unit test for empty 'params' in net45

### LINKED ISSUE #3272: HAA0101 should be reported for empty 'params' when targeting net45
💡 A new test needs to be added that uses the `net45` reference assemblies. These do not include `Array.Empty<T>()`, and the compiler may still insert an allocation.

_Originally posted by @sharwell in https://github.com/dotnet/roslyn-analyzers/pull/3270_

### PR DESCRIPTION
**Pull request by Evangelink**
Fix #3272

@sharwell is it what you were expecting?

### CODE REVIEW

**Comment by genlu** [comment]
Thanks! I think the actual issue is we are not checking for the existence of `Array.Empty<T>` in the analyzer, so in case of .net framework versions prior to 4.6, we would have false negative.

File: src/PerformanceSensitiveAnalyzers/UnitTests/CallSiteImplicitAllocationAnalyzerTests.cs [hunk 1]
```diff
@@ -147,5 +148,30 @@ public void SomeMethod()
...
+    public void Testing()
+    {
+        Params(); //no allocation
+    }
+
+    public void Params(params int[] args)
+    {
+    }
+}",
+            }.RunAsync();
```
Comment (genlu) [inline src/PerformanceSensitiveAnalyzers/UnitTests/CallSiteImplicitAllocationAnalyzerTests.cs]: Test code look good to me, just need to call though `VerifyCS.VerifyAnalyzerAsync` which will include attribute source.

File: src/PerformanceSensitiveAnalyzers/UnitTests/CallSiteImplicitAllocationAnalyzerTests.cs [hunk 2]
```diff
@@ -147,5 +148,30 @@ public void SomeMethod()
                 // Test0.cs(12,9): warning HAA0102: Non-overridden virtual method call on a value type adds a boxing or constrained instruction
                 VerifyCS.Diagnostic(CallSiteImplicitAllocationAnalyzer.ValueTypeNonOverridenCallRule).WithLocation(12, 9));
         }
+
+        [Fact, WorkItem(3272, "https://github.com/dotnet/roslyn-analyzers/issues/3272")]
+        public async Task C()
```
Comment (genlu) [inline src/PerformanceSensitiveAnalyzers/UnitTests/CallSiteImplicitAllocationAnalyzerTests.cs]: Test name

File: src/PerformanceSensitiveAnalyzers/UnitTests/CallSiteImplicitAllocationAnalyzerTests.cs [hunk 3]
```diff
@@ -147,5 +148,30 @@ public void SomeMethod()
...
+                TestCode = @"
+using System;
+using Roslyn.Utilities;
+
+public class MyClass
+{
+    [PerformanceSensitive(""uri"")]
+    public void Testing()
+    {
+        Params(); //no allocation
```
Comment (sharwell) [inline src/PerformanceSensitiveAnalyzers/UnitTests/CallSiteImplicitAllocationAnalyzerTests.cs]: ❗️ This is not correct. When targeting **net45** there _is_ an allocation here.

File: src/PerformanceSensitiveAnalyzers/UnitTests/CallSiteImplicitAllocationAnalyzerTests.cs (hunk 1 above)
Comment (sharwell) [reply src/PerformanceSensitiveAnalyzers/UnitTests/CallSiteImplicitAllocationAnalyzerTests.cs]: The full form is required since we are changing the reference assemblies.

File: src/PerformanceSensitiveAnalyzers/UnitTests/CallSiteImplicitAllocationAnalyzerTests.cs (hunk 2 above)
Comment (Evangelink) [reply src/PerformanceSensitiveAnalyzers/UnitTests/CallSiteImplicitAllocationAnalyzerTests.cs]: My bad, I went for a quick test and forgot to review... I will update!

File: src/PerformanceSensitiveAnalyzers/UnitTests/CallSiteImplicitAllocationAnalyzerTests.cs [hunk 4]
```diff
@@ -52,6 +53,44 @@ public void ParamsWithObjects(params object[] args)
...
+
+    public void Params(params int[] args)
+    {
+    }
+}",
+                        ("PerformanceSensitiveAttribute.cs", VerifyCS.PerformanceSensitiveAttributeSource)
+                    },
+                    ExpectedDiagnostics =
+                    {
+                        VerifyCS.Diagnostic(CallSiteImplicitAllocationAnalyzer.ParamsParameterRule).WithLocation(10, 9),
```
Comment (Evangelink) [inline src/PerformanceSensitiveAnalyzers/UnitTests/CallSiteImplicitAllocationAnalyzerTests.cs]: Alright, so now, the test is failing because no issue is raised. In the `CheckParam` method when debugging the net45 case and the other case I don't seem to see any difference:

https://github.com/dotnet/roslyn-analyzers/blob/master/src/PerformanceSensitiveAnalyzers/CSharp/CallSiteImplicitAllocationAnalyzer.cs#L79-L83

Comment (Evangelink) [reply src/PerformanceSensitiveAnalyzers/UnitTests/CallSiteImplicitAllocationAnalyzerTests.cs]: The only thing I can see (with my knowledge) is to access to the referenced assemblies through the compilation and to check the .net version using `mscorlib`. Then it that's earlier than net45 we do raise. Is it what you are thinking about?

Comment (sharwell) [reply src/PerformanceSensitiveAnalyzers/UnitTests/CallSiteImplicitAllocationAnalyzerTests.cs]: The allocation is only avoided if `System.Array.Empty<T>()` is defined.

**Comment by Evangelink** [comment]
Thank you @sharwell! You did basically all the job of the PR but hey happy to have "helped" :D

File: src/PerformanceSensitiveAnalyzers/CSharp/CallSiteImplicitAllocationAnalyzer.cs [hunk 5]
```diff
@@ -79,6 +79,11 @@ private static void CheckParam(InvocationExpressionSyntax invocationExpression,
             var arguments = invocationExpression.ArgumentList.Arguments;
             if (arguments.Count == methodInfo.Parameters.Length - 1)
             {
+                // Up to net45 the System.Array.Empty<T> singleton didn't existed so an empty params array was still causing some memory allocation.
+                if (semanticModel.Compilation.GetSpecialType(SpecialType.System_Array).GetMembers("Empty").Length == 0)
```
Comment (Evangelink) [inline src/PerformanceSensitiveAnalyzers/CSharp/CallSiteImplicitAllocationAnalyzer.cs]: I have tried to extract this to the `AnalyzeNode` method but this was actually causing a few more un-needed access. The best thing would be to have this check done once at the base-class level but I wasn't sure this was ok for you to update everything.

Comment (sharwell) [reply src/PerformanceSensitiveAnalyzers/CSharp/CallSiteImplicitAllocationAnalyzer.cs]: I'm OK with this for now since it's already gated on the attribute.

**Comment by Evangelink** [comment]
ping @genlu
